session/session.go

```diff
@@ -0,0 +1,94 @@
+// session implements a finality independent session system. It let session producers transparently integrate with any session consumer without prequired shared state.
+// A session bundles a logical set of related requests, it let download clients more efficiently send requests in the network, the assumption is that the same nodes should host most of the data in the related requests.
+// To create a new session use [ContextWithSession], compatible consumers like bitswap will automatically pick it up from the context.
+package session
+
+import (
+	"context"
+	"sync"
+)
+
+type contextKey struct{}
+
+// ContextWithSession attach a session to the context, it may reuse an existing session if one already exists in the context.
+// The context must be canceled in some way to ensure all sessions are cleaned up.
+func ContextWithSession(ctx context.Context) context.Context {
+	if _, ok := getRegistryFromContext(ctx); ok {
+		return ctx
+	}
+
+	return context.WithValue(ctx, contextKey{}, &sessionRegistry{sesCtx: ctx})
+}
+
+// sessionRegistry is meant for session consumers.
+// This type allows session consumers to inject their own session object within an existing context.
+// This exists as it allows to satisfy theses two requirements:
+// - Session producers must not need to hold some prexisting session factory object to inject into their contextes.
+// - Session consumers must be able to not to maintain any state to track inflight sessions, instead they can store their session object in the sessionRegistry and rely on GC to cleanup.
+// It is threadsafe.
+type sessionRegistry struct {
+	sesCtx   context.Context
```

Comment:
Why store this?

Reply:
For the same reason you originally did in your few years old PR.
Let's assume the consumer does this:
```go
sesCtx, cancel := context.WithCancel(context.Background())
defer cancel()

sesCtx = sessions.ContextWithSession(sesCtx)

var wg sync.WaitGroup
wg.Add(2)
go func() {
 defer wg.Done()
 ctx := context.WithTimeout(sesCtx, time.Second)
 useRequest1(bitswap.GetBlocks(ctx, request1))
}()
go func() {
 defer wg.Done()
 ctx := context.WithTimeout(sesCtx, time.Second*10)
 useRequest2(bitswap.GetBlocks(ctx, request2))
}()
wg.Wait()
```
Which ever one of the two request gets ran first will try to load the session from the session registry, it wont exists, create will be called and at this point within `GetBlocks` create need to use `sesCtx` not `ctx` because `GetBlocks` can have a timeout or cancel policy smaller than the whole session context.

If I passed the local context in the example above, if request one is scheduled first the whole session would be created with a one second timeout, then request two would have a timeout of one second too instead of the 10 it expects.

Reply:
```go
	ss, ok := create(r.sesCtx)
```